fdbserver/DataDistribution.actor.cpp

```diff
@@ -2490,6 +2536,65 @@ ACTOR Future<Void> removeBadTeams(DDTeamCollection* self) {
 	return Void();
 }
 
+bool inCorrectDC(DDTeamCollection* self, TCServerInfo* server) {
+	return (self->includedDCs.empty() ||
+	        std::find(self->includedDCs.begin(), self->includedDCs.end(), server->lastKnownInterface.locality.dcId()) !=
+	            self->includedDCs.end());
+}
+
+ACTOR Future<Void> removeWrongStoreType(DDTeamCollection* self) {
+	state int numServersRemoved = 0;
+	state std::map<UID, Reference<TCServerInfo>>::iterator server;
+	state vector<Reference<TCServerInfo>> serversToRemove;
+	state int i = 0;
+
+	loop {
+		wait(delay(1.0));
+		TraceEvent("WrongStoreTypeRemoverStartLoop", self->distributorId)
+		    .detail("Primary", self->primary)
+		    .detail("ServerInfoSize", self->server_info.size())
+		    .detail("SysRestoreType", self->configuration.storageServerStoreType);
+		serversToRemove.clear();
+		for (server = self->server_info.begin(); server != self->server_info.end(); ++server) {
+			NetworkAddress a = server->second->lastKnownInterface.address();
+			AddressExclusion addr(a.ip, a.port);
+			TraceEvent("WrongStoreTypeRemover", self->distributorId)
+			    .detail("DDID", self->distributorId)
+			    .detail("Server", server->first)
+			    .detail("Addr", addr.toString())
+			    .detail("StoreType", server->second->storeType)
+			    .detail("IsCorrectStoreType",
+			            server->second->isCorrectStoreType(self->configuration.storageServerStoreType))
+			    .detail("ToRemove", server->second->toRemove);
+			if (!server->second->isCorrectStoreType(self->configuration.storageServerStoreType) ||
+			    !inCorrectDC(self, server->second.getPtr())) {
+				serversToRemove.push_back(server->second);
+			} else {
+				server->second->toRemove =
+				    0; // In case the configuration.storeType is changed back to the server's type
+			}
+		}
+
+		for (i = 0; i < serversToRemove.size(); i++) {
+			Reference<TCServerInfo> s = serversToRemove[i];
+			if (s.isValid()) {
+				s->toRemove++; // The server's location will not be excluded
+				s->wrongStoreTypeRemoved.trigger();
```

Comment:
make wrongStoreTypeRemoved an asyncVar<bool> so you always know if it should be removed or not